File: 20-actors/uchiwake/methods/uchiwake_edn.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/uchiwake_edn.py (replaces the failed unit_refactor stub).
(ns uchiwake.methods.uchiwake-edn
  "uchiwake 内訳 — shared EDN reader + datom classifier + GTIN helpers (ADR-2606081800).

  The Python source ships a hand-rolled minimal EDN reader because Python has no EDN. Clojure
  reads EDN natively (`clojure.edn/read-string`), so the reader collapses to that — keyword
  keys (:product/id …) and keyword values (:metal, :representative) are read as real keywords
  (the Python port kept them as \":ns/name\" strings). The classifier + GTIN check-digit
  validation are ported faithfully so the analysis is byte-for-byte equivalent."
  (:require [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(defn load-edn [path] (edn/read-string (slurp (io/file path))))

(defn classify
  "Bucket a flat datom vector into uchiwake entity kinds. products/parts/materials are id→node
  maps; the rest are vectors (insertion order preserved)."
  [rows]
  (reduce
   (fn [out r]
     (if-not (map? r)
       out
       (cond
         (:product/id r)          (assoc-in out [:products (:product/id r)] r)
         (:part/id r)             (assoc-in out [:parts (:part/id r)] r)
         (:material/id r)         (assoc-in out [:materials (:material/id r)] r)
         (:bom.edge/id r)         (update out :bom conj r)
         (:process.step/id r)     (update out :process conj r)
         (:logistics.leg/id r)    (update out :logistics conj r)
         (:design.ref/id r)       (update out :design conj r)
         (:company.ownership/id r) (update out :ownership conj r)
         :else out)))
   {:products {} :parts {} :materials {}
    :bom [] :process [] :logistics [] :design [] :ownership []}
   rows))

;; ── GTIN helpers ──────────────────────────────────────────────────────────────
(defn- digits [gtin] (apply str (filter #(Character/isDigit ^char %) (str gtin))))

(defn normalize-gtin
  "Left-zero-pad any GTIN-8/12/13 to the canonical 14-digit GTIN-14."
  [gtin]
  (let [d (digits gtin)]
    (str (apply str (repeat (max 0 (- 14 (count d))) \0)) d)))

(defn gtin-check-digit-ok?
  "Validate the GS1 mod-10 check digit of a GTIN (length 8/12/13/14)."
  [gtin]
  (let [d (digits gtin)]
    (if-not (#{8 12 13 14} (count d))
      false
      (let [body (butlast d)
            check (Character/digit ^char (last d) 10)
            total (reduce + (map-indexed
                             (fn [i ch] (* (Character/digit ^char ch 10) (if (even? i) 3 1)))
                             (reverse body)))]
        (= (mod (- 10 (mod total 10)) 10) check)))))

File: 20-actors/uchiwake/methods/uchiwake_edn.py
"""uchiwake 内訳 — shared minimal EDN reader + datom classifier (stdlib only).

Ported from the kabuto/watatsuna readers (same subset: vectors [], maps {},
:keyword, "string", number, bool, nil). Keeps uchiwake's cells dependency-free so
they run on any python3 with no install step. ADR-2606081800.
"""
from __future__ import annotations
import re
import pathlib

# ── minimal EDN reader (subset) ──────────────────────────────────────────────
_TOK = re.compile(r'[\s,]+|;[^\n]*|(\[|\]|\{|\}|"(?:\\.|[^"\\])*"|[^\s,\[\]{}]+)')
_END = object()


def _tokens(s: str):
    for m in _TOK.finditer(s):
        t = m.group(1)
        if t is not None:
            yield t


def _atom(t: str):
    if t.startswith('"'):
        return t[1:-1].replace('\\"', '"').replace('\\\\', '\\')
    if t == 'true':
        return True
    if t == 'false':
        return False
    if t == 'nil':
        return None
    if t.startswith(':'):
        return t  # keep keywords as ":ns/name" strings
    try:
        return int(t)
    except ValueError:
        try:
            return float(t)
        except ValueError:
            return t


def _parse(it):
    t = next(it)
    if t == '[':
        out = []
        while (x := _parse(it)) is not _END:
            out.append(x)
        return out
    if t == '{':
        out = {}
        while (k := _parse(it)) is not _END:
            v = _parse(it)
            out[k] = v
        return out
    if t in (']', '}'):
        return _END
    return _atom(t)


def load_edn(path: pathlib.Path):
    it = _tokens(pathlib.Path(path).read_text(encoding='utf-8'))
    return _parse(it)


def read_str(text: str):
    """Parse the first top-level form from EDN text (1:1 with read-edn in the .cljc port).
    Used to read one append-only kotoba Datom-log line back (one tx map per line)."""
    return _parse(_tokens(text))


# ── classify the flat datom vector into entity buckets ───────────────────────
def classify(rows):
    """Return a dict of entity buckets keyed/listed by uchiwake entity kind."""
    out = {
        'products': {}, 'parts': {}, 'materials': {},
        'bom': [], 'process': [], 'logistics': [], 'design': [], 'ownership': [],
    }
    for r in rows:
        if not isinstance(r, dict):
            continue
        if ':product/id' in r:
            out['products'][r[':product/id']] = r
        elif ':part/id' in r:
            out['parts'][r[':part/id']] = r
        elif ':material/id' in r:
            out['materials'][r[':material/id']] = r
        elif ':bom.edge/id' in r:
            out['bom'].append(r)
        elif ':process.step/id' in r:
            out['process'].append(r)
        elif ':logistics.leg/id' in r:
            out['logistics'].append(r)
        elif ':design.ref/id' in r:
            out['design'].append(r)
        elif ':company.ownership/id' in r:
            out['ownership'].append(r)
    return out


def edn_str(s: str) -> str:
    """EDN-escape a python string into a quoted EDN string literal."""
    return '"' + str(s).replace('\\', '\\\\').replace('"', '\\"') + '"'


# ── GTIN helpers ─────────────────────────────────────────────────────────────
def normalize_gtin(gtin: str) -> str:
    """Left-zero-pad any GTIN-8/12/13 to the canonical 14-digit GTIN-14."""
    d = ''.join(ch for ch in str(gtin) if ch.isdigit())
    return d.zfill(14)


def gtin_check_digit_ok(gtin: str) -> bool:
    """Validate the GS1 mod-10 check digit of a GTIN (any length 8/12/13/14)."""
    d = ''.join(ch for ch in str(gtin) if ch.isdigit())
    if len(d) not in (8, 12, 13, 14):
        return False
    body, check = d[:-1], int(d[-1])
    # GS1: rightmost body digit weighted x3, alternating.
    total = 0
    for i, ch in enumerate(reversed(body)):
        total += int(ch) * (3 if i % 2 == 0 else 1)
    return (10 - (total % 10)) % 10 == check
